File: Drivers and Simulators/FilterWheel Simulator .NET/FilterWheelDriver/PictureBoxArray.vb
<ComVisible(False)> _
Public Class PictureBoxArray

    Inherits System.Collections.CollectionBase
    Private ReadOnly HostForm As System.Windows.Forms.Form


    Public Function AddNewPictureBox() As System.Windows.Forms.PictureBox
        ' Create a new instance of the Button class.
        Dim aPictureBox As New System.Windows.Forms.PictureBox()
        Dim aTooltip As New ToolTip()
        ' Add the button to the collection's internal list.
        Me.List.Add(aPictureBox)
        ' Add the button to the controls collection of the form 
        ' referenced by the HostForm field.
        HostForm.Controls.Add(aPictureBox)
        ' Set intial properties for the button object.
        aPictureBox.Top = Count * 25
        aPictureBox.Left = 100
        aPictureBox.BorderStyle = BorderStyle.FixedSingle
        aPictureBox.Tag = Me.Count
        aPictureBox.BackColor = Drawing.Color.DarkGray
        aTooltip.SetToolTip(aPictureBox, "Click to select colour")

        AddHandler aPictureBox.Click, AddressOf ClickHandler

        Return aPictureBox
    End Function

    Public Sub New(ByVal host As System.Windows.Forms.Form)
        HostForm = host
        'Me.AddNewPictureBox()
    End Sub

    Default Public ReadOnly Property Item(ByVal Index As Integer) As System.Windows.Forms.PictureBox
        Get
            Return CType(Me.List.Item(Index), System.Windows.Forms.PictureBox)
        End Get
    End Property

    Public Sub Remove()
        ' Check to be sure there is a PictureBox to remove.
        If Me.Count > 0 Then
            ' Remove the last button added to the array from the host form 
            ' controls collection. Note the use of the default property in 
            ' accessing the array.
            HostForm.Controls.Remove(Me(Me.Count - 1))
            Me.List.RemoveAt(Me.Count - 1)
        End If
    End Sub

    ' When the picturebox is clicked show a colour picker dialog
    ' set initially to the current colour. If OK clicked on colour picker
    ' set the current picture box background to the selected colour
    Public Sub ClickHandler(ByVal sender As Object, ByVal e As System.EventArgs)
        Dim colourDiag As Windows.Forms.ColorDialog = New Windows.Forms.ColorDialog
        colourDiag.AllowFullOpen = True
        colourDiag.AnyColor = True
        colourDiag.Color = CType(sender, System.Windows.Forms.PictureBox).BackColor
        If (colourDiag.ShowDialog() = DialogResult.OK) Then _
            CType(sender, System.Windows.Forms.PictureBox).BackColor = colourDiag.Color
    End Sub


End Class

File: tags/Build_5.1.315.Weekly/Drivers and Simulators/FilterWheel Simulator .NET/FilterWheelServer/PictureBoxArray.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Drawing;

namespace ASCOM.FilterWheelSim
{
    class PictureBoxArray : System.Collections.CollectionBase
    {
        private System.Windows.Forms.Form HostForm;

        public PictureBoxArray(Form host)
        {
            HostForm = host;
            //Me.AddNewPictureBox()
        }

        public PictureBox AddNewPictureBox()
        {
            // Create a new instance of the PictureBox class.
            PictureBox aPictureBox = new PictureBox();
            ToolTip aTooltip = new ToolTip();
            // Add the PictureBox to the collection's internal list.
            this.List.Add(aPictureBox);
            // Add the PictureBox to the controls collection of the form 
            // referenced by the HostForm field.
            HostForm.Controls.Add(aPictureBox);
            // Set intial properties for the PictureBox object.
            aPictureBox.Top = this.Count * 25;
            aPictureBox.Left = 100;
            aPictureBox.BorderStyle = BorderStyle.FixedSingle;
            aPictureBox.Tag = this.Count;
            aPictureBox.BackColor = Color.DarkGray;
            aTooltip.SetToolTip(aPictureBox, "Click to select colour");

            aPictureBox.Click += new System.EventHandler(ClickHandler);

            return aPictureBox;
        }

         public PictureBox this[int Index]
        {
            get { return (PictureBox)this.List[Index]; }
        }

        public void Remove()
        {
            // Check to be sure there is a PictureBox to remove.
            if (this.Count >= 1)
            {
                // Remove the last PictureBox added to the array from the host form 
                // controls collection. Note the use of the default property in 
                // accessing the array.
                HostForm.Controls.Remove(this[this.Count - 1]);
                this.List.RemoveAt(this.Count - 1);
            }
        }

        // When the picturebox is clicked show a colour picker dialog
        // set initially to the current colour. If OK clicked on colour picker
        // set the current picture box background to the selected colour
        public void ClickHandler(Object sender, System.EventArgs e)
        {
            ColorDialog colourDiag = new ColorDialog();
            colourDiag.AllowFullOpen = true;
            colourDiag.AnyColor = true;
            colourDiag.Color = ((PictureBox)sender).BackColor;
            if (colourDiag.ShowDialog() == DialogResult.OK)
            ((PictureBox)sender).BackColor = colourDiag.Color;
        }

    }
}
